includes/Core/Util/Migration_1_8_1.php

```diff
@@ -0,0 +1,262 @@
+
+use Google\Site_Kit\Context;
+use Google\Site_Kit\Core\Authentication\Authentication;
+use Google\Site_Kit\Core\Authentication\Google_Proxy;
+use Google\Site_Kit\Core\Authentication\Profile;
+use Google\Site_Kit\Core\Authentication\Verification_File;
+use Google\Site_Kit\Core\Authentication\Verification_Meta;
+use Google\Site_Kit\Core\Permissions\Permissions;
+use Google\Site_Kit\Core\Storage\Options;
+use Google\Site_Kit\Core\Storage\User_Options;
+use WP_User;
+use WP_Error;
+
+/**
+ * Class Migration_1_8_1
+ *
+ * @since 1.8.1
+ * @access private
+ * @ignore
+ */
+class Migration_1_8_1 {
+	/**
+	 * Target DB version.
+	 */
+	const DB_VERSION = '1.8.1';
+
+	/**
+	 * Context instance.
+	 *
+	 * @since 1.8.1
+	 * @var Context
+	 */
+	protected $context;
+
+	/**
+	 * Options instance.
+	 *
+	 * @since 1.8.1
+	 * @var Options
+	 */
+	protected $options;
+
+	/**
+	 * User_Options instance.
+	 *
+	 * @since 1.8.1
+	 * @var User_Options
+	 */
+	protected $user_options;
+
+	/**
+	 * Authentication instance.
+	 *
+	 * @since 1.8.1
+	 * @var Authentication
+	 */
+	protected $authentication;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since 1.8.1
+	 *
+	 * @param Context        $context        Plugin context instance.
+	 * @param Options        $options        Optional. Options instance.
+	 * @param User_Options   $user_options   Optional. User_Options instance.
+	 * @param Authentication $authentication Optional. Authentication instance. Default is a new instance.
+	 */
+	public function __construct(
+		Context $context,
+		Options $options = null,
+		User_Options $user_options = null,
+		Authentication $authentication = null
+	) {
+		$this->context        = $context;
+		$this->options        = $options ?: new Options( $this->context );
+		$this->user_options   = $user_options ?: new User_Options( $this->context );
+		$this->authentication = $authentication ?: new Authentication( $this->context, $this->options, $this->user_options );
+	}
+
+	/**
+	 * Registers hooks.
+	 *
+	 * @since 1.8.1
+	 */
+	public function register() {
+		add_action( 'admin_init', array( $this, 'migrate' ) );
+	}
+
+	/**
+	 * Migrates the DB.
+	 *
+	 * @since 1.8.1
+	 *
+	 * @return boolean|WP_Error True on success, WP_Error on failure, false if
+	 *                          routine cannot run or is not applicable.
+	 */
+	public function migrate() {
+		$db_version = $this->options->get( 'googlesitekit_db_version' );
+
+		// Do not run if database version already updated.
+		if ( $db_version && version_compare( $db_version, self::DB_VERSION, '>=' ) ) {
+			return false;
+		}
+
+		// Only run routine if using the authentication service, otherwise it
+		// is irrelevant.
+		if ( ! $this->authentication->get_oauth_client()->using_proxy() ) {
+			return false;
+		}
+
+		// Only run routine once site credentials present, otherwise it is not
+		// possible to connect to the authentication service.
+		if ( ! $this->authentication->credentials()->has() ) {
+			return false;
+		}
```

Comment:
Yeah I took that into consideration, I think it's acceptable here. I think we should ensure we wait with the routine until the site is connected to the proxy. The checks that are made here aren't expensive, so I think only updating the DB version / running the routine once these requirements are met makes sense.

We shouldn't decouple the user detection/cleanup from the proxy API call since it needs to send related information there.